core/src/invalid_fee_payer_filter.rs

```diff
@@ -0,0 +1,108 @@
+//! Manages a set of recent invalid fee payers.
+//! This is used to filter out transactions that are unlikely to be included
+//! in any block, much earlier in the transaction pipeline.
+//!
+
+use {
+    dashmap::DashSet,
+    solana_sdk::pubkey::Pubkey,
+    std::{
+        sync::{
+            atomic::{AtomicBool, AtomicUsize, Ordering},
+            Arc,
+        },
+        thread::{sleep, Builder, JoinHandle},
+        time::Duration,
+    },
+};
+
+#[derive(Default)]
+pub struct InvalidFeePayerFilter {
+    recent_invalid_fee_payers: DashSet<Pubkey>,
```

Comment:
Had that in mind as a follow up, hence why I separated a class.
I can just do in this PR tho

Reply:
Looked into this a bit and could be wrong. It looks to me reseting or clearing the deduper/bloom filter would require mutability; so we'd need some sort of locking mechanism. Since this filter is shared by multiple threads, that would block other threads which is something I'm trying to avoid.

Might be better to just vacate LRU if we would exceed some capacity. That would protect us from OOMing, but potentially slow things down if we exceed capacity. vs periodic locking of something like deduper. wdyt?

Reply:
isn't deduper all atomics?

Reply:
Not for resetting:

```rust
    pub fn maybe_reset<R: Rng>(
        &mut self,
        ...
```

because the underlying `bits` is stored in a `Vec`: `    bits: Vec<AtomicU64>,`

Reply:
Played around with this a bit in https://github.com/apfitzge/solana/tree/invalid_fee_payer_filter_bloom, where I atomically swapped between 2 bloom filters so we were (mostly) not stepping on our feet.


However, I'm really hesitant to add in a probablistic data struct for banking stage which would cause us to not process valid packets. I think that's generally fine for runtime as the false positive rate is low, but I think it'd make many of our banking tests flaky.

I think the main concern is resource exhaustion. I opted for a very simple random eviction from the dashmap upon exceeding capacity so we don't make our tests flaky.

Reply:
performance in the bench i was running (1 very active invalid fee-payer) showed no difference between bloom and dash map internally.

Reply:
@t-nelson are you content with this decision to just randomly evict if we see too many invalid fee payers?

Reply:
> seems like bloom should be preferable in that case due to constant size?

thoughts?

Reply:
a pre-allocated dashmap also has a constant size, although I think larger than the double-bloom filter for a similar number of items.

Since there doesn't seem to be any **time** performance benefit, I was sticking with the simpler implementation, instead of maintaining 2 bloom-filters to allow us to "atomically" clear.